semgrep-core/src/parsing/Parse_rule.ml

```diff
@@ -43,64 +49,107 @@ module H = Parse_mini_rule
 (* Helpers *)
 (*****************************************************************************)
 
-(* less: could use a hash to accelerate things *)
-let rec find_fields flds xs =
-  match flds with
-  | [] -> ([], xs)
-  | fld :: flds ->
-      let fld_match = List.assoc_opt fld xs in
-      let xs = List.remove_assoc fld xs in
-      let matches, rest = find_fields flds xs in
-      ((fld, fld_match) :: matches, rest)
+type env = { id : string; languages : R.xlang; path : string list }
 
 let error s = raise (E.InvalidYamlException s)
 
-let rec yaml_to_json = function
-  | `Null -> J.Null
-  | `Bool b -> J.Bool b
-  | `Float f ->
-      (* or use J.Int? *)
-      J.Float f
-  | `String s -> J.String s
-  | `A xs -> J.Array (xs |> List.map yaml_to_json)
-  | `O xs -> J.Object (xs |> List.map (fun (k, v) -> (k, yaml_to_json v)))
+let generic_to_json key ast =
+  let rec generic_to_json_help = function
+    | G.L (Null _) -> J.Null
+    | G.L (Bool (b, _)) -> J.Bool b
+    | G.L (Float (Some f, _)) -> J.Float f
+    | G.L (Int (Some i, _)) -> J.Int i
+    | G.L (String (s, _)) -> J.String s
+    | G.Container (Array, (_, xs, _)) ->
+        J.Array (xs |> List.map generic_to_json_help)
+    | G.Container (Dict, (_, xs, _)) ->
+        J.Object
+          (xs
+          |> List.map (fun x ->
+                 match x with
+                 | G.Tuple (_, [ L (String (k, _)); v ], _) ->
+                     (k, generic_to_json_help v)
+                 | _ -> error ("Expected key value pair in " ^ key ^ " dict")))
+    | _ -> error "Unexpected generic representation of yaml"
+  in
+  generic_to_json_help ast
+
+let opt_to_list = function None -> [] | Some xs -> xs
+
+(*****************************************************************************)
+(* Dict helper methods *)
+(*****************************************************************************)
+
+let yaml_to_dict name rule =
+  match rule with
+  | G.Container (Dict, (_, fields, _)) ->
+      let dict = Hashtbl.create 10 in
+      fields
+      |> List.iter (fun field ->
+             match field with
+             | G.Tuple (_, [ L (String (key, _)); value ], _) ->
+                 Hashtbl.add dict key value
+             | _ -> error "Not a valid key value pair");
+      dict
+  | _ -> error ("each " ^ name ^ " should be a dictionary of fields")
```

Comment:
Perhaps the error message should talk about "object" (YAML/JSON terminology) rather than "dictionary of fields" (the way we chose map a JSON object to the Generic AST)?

Reply:
done